tests/41end-to-end-keys/08-cross-signing.pl

```diff
@@ -682,7 +682,17 @@
       })->then( sub {
          sync_until_user_in_device_list( $user1, $user2 );
       })->then( sub {
-         matrix_get_e2e_keys( $user1, $user2_id );
+         repeat_until_true {
+            # Wait until user1 sees signatures uploaded by user2.
+            # We repeat_until_true because this is sensitive to replication delays.
+            matrix_get_e2e_keys( $user1, $user2_id )
+            ->then(sub {
+               my ( $content ) = @_;
+               if ( exists $content->{device_keys}->{$user2_id}->{$user2_device}->{"signatures"}) {
+                  return $content;
```

Comment:
From what I'm seeing of other instances of `repeat_until_true`, one typically calls `Future->done( <true || false> )` to retry or complete the loop. Does `return` do the same thing here?

Reply:
> From what I'm seeing of other instances of `repeat_until_true`, one typically calls `Future->done( <true || false> )` to retry or complete the loop. Does `return` do the same thing here?

I think those examples will probably have `Future->done( <true || false> )` as their last expression---which is a Rust-like implicit return. Quoting https://perldoc.perl.org/perlsub :

> If no`return` is found and if the last statement is an expression, its value is returned.

I think what I actually wanted to use here was `retry_until_success`, which has an upper bound on the number of retry attempts.

Reply:
In a little more detail:

- repeat_until_true repeatedly calls a function `f` which returns a future.
- If `f` returns a future that is `done`, We inspect its encapsulated value. If that value is true, we stop the loop.
- Otherwise we continue.
So the call to `Future->done( <true || false> )` that's special; it's the fact that we're returning it. (Think of it as a Rustic `Result<bool, Error>` holding `true` or `false`---it's just a lump of data)

Reply:
I see, and yes I forgot about that Perl quirk - thanks!

By writing `return $content;`, are we returning a `Future` though? To me it just looks like we're returning a dictionary?

Reply:
I'm with anoa on this - I thought that functions passed to `then` had to return a `Future` rather than an immediate value (whether that be by calling another `Future`-returning function, or by creating a completed `Future` with `Future->done()`).

It's possible that requirement has been relaxed, however.

Reply:
(and I don't think this is specific to `repeat_until_true` or `retry_until_success`).

Reply:
I think you're both right. Quoting https://metacpan.org/pod/Future#then:

Returns a new sequencing Future that runs the code if the first succeeds. Once $f1 succeeds the code reference will be invoked and is passed the list of results. **It should return a future, $f2.** Once $f2 completes the sequence future will then be marked as complete with whatever result $f2 gave. If $f1 fails then the sequence future will immediately fail with the same failure and the code will not be invoked.

So I have no idea how this worked.

In any case, c820c95ef4f968a20a2390e544b28842e91d188c makes sure that the argument to `then` returns a `Future` or `die`s.

Reply:
a bit further up:

> Since version 0.45: if a non-future result is returned it will be wrapped in a new immediate Future instance.

Reply:
/me gains a new appreciation for `python -m this | grep obvious`